src/astral/data-utils.js

```diff
@@ -95,6 +95,51 @@ const getCharacter = async (name) => {
     return char ? char.id : undefined;
 }
 
+const getCharacterData = async (id) => await (await fetch(
+    `${location.origin}/api/game/${getRoom()}/character/${id}`, 
+    {
+        method: "GET", 
+        headers: await getHeaders()
+    }
+)).json();
+
+const updateCustomAttribute = async (characterData, attrbiuteName, value) => {
+    let attr = characterData.customAttributes.find(attr => attr.name == attrbiuteName);
+    if (!attr) {
+        attr = {
+            name: attrbiuteName,
+            value: value.toString()
+        }
+        characterData.customAttributes.push(attr);
+        return;
+    }
+    attr.value = value.toString();
+}
+
+const updateResourceBar = async (characterData, barName, value, maxValue, color = null, actualValue = null, actualMaxValue = null, display = true) => {
+    let bar = characterData.resourceBars.find(attr => attr.label == barName);
+    if (!bar) {
+        bar = {
+            label: barName,
+            system: false,
+            color: color ? Number.parseInt(color, 16) : Math.floor(Math.random()*16777215)
+        }
+        characterData.resourceBars.push(bar);
+    }
+    Object.assign(bar, {
+        leftEquation: value.toString(),
+        rightEquation: maxValue.toString(),
+        display,
+        ...(actualValue != null ? {
+            leftValue: actualValue
+        } : {}),
+        ...(actualMaxValue  != null ? {
+            rightValue: actualMaxValue
+        } : {}),
+        ...(color && !bar.color ? { color } : {})
```

Comment:
Same comment as above (or below?) with the fetch. The construct works, but makes it hard for readability/maintainability. Better to spread it out a bit in my opinion.